src/v/kafka/security/scram_algorithm.cc

```diff
@@ -0,0 +1,205 @@
+// clang-format on
+
+std::ostream& operator<<(std::ostream& os, const scram_credential& cred) {
+    fmt::print(
+      os,
+      "salt {} server_key {} stored_key {} iterations {}",
+      cred._salt,
+      cred._server_key,
+      cred._stored_key,
+      cred._iterations);
+    return os;
+}
+
+client_first_message::client_first_message(bytes_view data) {
+    static const std::regex re(
+      client_first_message_re, std::regex::ECMAScript | std::regex::optimize);
+
+    // std::regex really likes using std::string
+    std::string message(data.begin(), data.end());
+    validate_utf8(message);
+
+    std::smatch match;
+    if (unlikely(!std::regex_match(message, match, re))) {
+        throw scram_exception(
+          fmt::format("Invalid SCRAM client first message: {}", message));
+    }
+
+    _authzid = match[1].str();
+    _username = match[3].str();
+    _nonce = match[4].str();
+
+    auto extensions = match[5].str(); // NOLINT
+    if (extensions.empty()) {
+        return;
+    }
+
+    // split on "," following the "," prefix
+    std::vector<std::string> extension_pairs;
+    boost::split(
+      extension_pairs, extensions.substr(1), [](char c) { return c == ','; });
+
+    // split pairs on first "=". the value part may also contain "="
+    for (const auto& pair : extension_pairs) {
+        auto it = std::find(pair.cbegin(), pair.cend(), '=');
+        if (unlikely(it == pair.cend())) {
+            throw scram_exception(
+              fmt::format("Invalid SCRAM client first message: {}", message));
+        }
+        _extensions.emplace(
+          ss::sstring(pair.cbegin(), it), ss::sstring(it + 1, pair.cend()));
+    }
+}
+
+ss::sstring client_first_message::username_normalized() const {
+    static const std::regex comma_re(
+      "=2C", std::regex::ECMAScript | std::regex::optimize);
+
+    static const std::regex equal_re(
+      "=3D", std::regex::ECMAScript | std::regex::optimize);
+
+    // first replace =2C with ','
+    auto normalized = std::regex_replace(_username.c_str(), comma_re, ",");
+
+    // check that after replacing =3D with '=' there are no other '='
+    if (unlikely(
+          std::regex_replace(normalized, equal_re, "").find('=')
+          != std::string::npos)) {
+        throw scram_exception(
+          fmt::format("Invalid SCRAM username: {}", _username));
+    }
+
+    return std::regex_replace(normalized, equal_re, "=");
+}
+
+bool client_first_message::token_authenticated() const {
+    if (auto it = _extensions.find("tokenauth"); it != _extensions.end()) {
+        return boost::iequals(it->second, "true");
+    }
+    return false;
+}
+
+std::ostream& operator<<(std::ostream& os, const client_first_message& msg) {
+    fmt::print(
+      os,
+      "authzid {} username {} nonce {}",
+      msg._authzid,
+      msg._username,
+      msg._nonce);
+    return os;
+}
+
+ss::sstring server_first_message::sasl_message() const {
```

Comment:
The conversions to `ss::string` create copies. I wonder if we should customise `fmt` to support `ss::string` at some point? Or use a stream interface? It might not matter all that much here, I suspect the messages are quite small.

Reply:
they are small but this would be a cool thing to do